Lib/glob.py

```diff
@@ -310,3 +157,229 @@ def translate(pat, *, recursive=False, include_hidden=False, seps=None):
+    while parts and magic_check.search(parts[0]) is None:
+        # Consume next non-wildcard component (speeds up joining).
+        if parts[0] not in _special_parts:
+            is_special = False
+        part += os.path.sep + parts[0]
+        parts = parts[1:]
+
+    select_next = _selector(parts, recursive, include_hidden)
+
+    def select_literal(path, rel_path, dir_fd, exists):
+        path = _add_trailing_slash(path) + part
+        rel_path = _add_trailing_slash(rel_path) + part
+        yield from select_next(path, rel_path, dir_fd, exists and is_special)
+    return select_literal
+
+
+def _wildcard_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects direct children of a given path,
+    filtering by pattern.
+    """
+    if include_hidden and part == '*':
+        match = None  # Skip generating a pattern that would match all inputs.
+    else:
+        match = _compile_pattern(part, recursive, include_hidden)
+
+    if parts:
+        select_next = _selector(parts, recursive, include_hidden)
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                if fd is not None:
+                    prefix = _add_trailing_slash(path)
+                with os.scandir(arg) as scandir_it:
+                    entries = list(scandir_it)
+                for entry in entries:
+                    if match is None or match(entry.name):
+                        try:
+                            if entry.is_dir():
+                                entry_path = entry.path
+                                if fd is not None:
+                                    entry_path = prefix + entry_path
+                                yield from select_next(entry_path, entry.name, fd, True)
+                        except OSError:
+                            pass
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    else:
+        def select_wildcard(path, rel_path, dir_fd, exists):
+            close_fd = False
+            try:
+                arg, fd, close_fd = _open_dir(path, rel_path, dir_fd)
+                prefix = _add_trailing_slash(path)
+                for name in os.listdir(arg):
+                    if match is None or match(name):
+                        yield prefix + name
+            except OSError:
+                pass
+            finally:
+                if close_fd:
+                    os.close(fd)
+
+    return select_wildcard
+
+
+def _recursive_selector(part, parts, recursive, include_hidden):
+    """Returns a function that selects a given path and all its children,
+    recursively, filtering by pattern.
+    """
+    while parts and parts[0] == '**':
+        parts = parts[1:]
+    while parts and parts[0] not in _special_parts:
```

Comment:
Same as above, a few suggestions:
```python
part += os.path.sep + os.path.join(itertools.takewhile(lambda x: x not in _special_parts, parts))
```
or
```python
while ipart := (itertools.takewhile(lambda x: x not in _special_parts, parts)):
    part += os.path.sep + ipart
```
whichever is more readable

Reply:
What's the advantage of those snippets vs the original code?

Reply:
This part of the code doesn't need to be particularly optimized, because it's only run when parsing a pattern (not on every path visited), and the result is cached. The pattern `parts` tuple is usually small because glob patterns are usually small, like `('**', '*', '*.py')`.

Reply:
Same as above, put as you see fit :)

Reply:
```python
def method5(parts):
     return os.path.sep + os.path.sep.join([*itertools.takewhile(lambda x: x not in _special_parts, parts)])
```

method5 is as fast as method 4, I wanted to check the performance cost of list unpacking and it doesn't seem to affect all that much.